### PR TITLE
Documentation for Hypothermia System

### PR DESCRIPTION
**Pull request by ktksan**

### CODE REVIEW

File: src/main/java/org/terasology/climateConditions/VisibleBreathComponent.java [hunk 1]
```diff
@@ -3,6 +3,10 @@
 import org.terasology.entitySystem.Component;
 import org.terasology.entitySystem.entity.EntityRef;
 
+/**
+ * This component is added to the player entity when the player enters the danger zone. Gives a visible breath to the
+ * player using particle effects.
+ */
 public class VisibleBreathComponent implements Component {
     public EntityRef particleEntity = EntityRef.NULL;
 }
```
Comment (jdrueckert) [inline src/main/java/org/terasology/climateConditions/VisibleBreathComponent.java]:
```suggestion
}

```

File: src/main/java/org/terasology/climateConditions/FrostbiteSystem.java [hunk 2]
```diff
@@ -36,6 +36,10 @@
 import org.terasology.utilities.random.FastRandom;
 import org.terasology.utilities.random.Random;
 
+/**
+ * This system is responsible for periodic frostbite damage (along with stun and a sound effect) to the player as long
+ * as the player has a Hypothermia component.
+ */
```
Comment (jdrueckert) [inline src/main/java/org/terasology/climateConditions/FrostbiteSystem.java]:
```suggestion
/**
 * Adds frostbite to the player.
 * Frostbite is a periodic effect that, e.g., damages and stuns the player.
 * Is only active iff the player has a {@link HypothermiaComponent}.
 */
```

With "iff" being "if and only if".

File: src/main/java/org/terasology/climateConditions/HypothermiaComponent.java [hunk 3]
```diff
@@ -17,5 +17,8 @@
 
 import org.terasology.entitySystem.Component;
 
+/**
+ * This component is added by the authority when a player enters the danger zone (extreme cold climate).
+ */
```
Comment (jdrueckert) [inline src/main/java/org/terasology/climateConditions/HypothermiaComponent.java]:
```suggestion
/**
 * Increases the game difficulty in locations with extreme cold climate.
 * Is added/removed by the {@link HypothermiaSystem} when the player enters/leaves a "danger zone".
 */
```

File: src/main/java/org/terasology/climateConditions/HypothermiaSystem.java [hunk 4]
```diff
@@ -27,6 +27,11 @@
 import org.terasology.logic.players.PlayerCharacterComponent;
 import org.terasology.physics.events.MovedEvent;
 
+/**
+ * This system is responsible for adding a Hypothermia Component to the player whenever a player enters the danger
+ * zone(extreme cold climate). And also for modifying the speed of the player as long as the player has a Hypothermia
+ * Component.
+ */
```
Comment (jdrueckert) [inline src/main/java/org/terasology/climateConditions/HypothermiaSystem.java]:
```suggestion
/**
 * Adds a {@link HypothermiaComponent} to the player.
 * Hypothermia occurs in locations with extremely cold climate and, e.g., slows the player's movements.
 */
```

File: src/main/java/org/terasology/climateConditions/HypothermiaSystem.java [hunk 5]
```diff
@@ -48,13 +53,19 @@ public void observeDangerZone(MovedEvent event, EntityRef player) {
         }
     }
 
+    /**
+     * Modifies the walking/running speed of player as long as the player has a Hypothermia Component.
+     */
```
Comment (jdrueckert) [inline src/main/java/org/terasology/climateConditions/HypothermiaSystem.java]:
```suggestion
    /**
     * Modifies the walking/running speed of the player.
     * Is only active iff the player has a {@link HypothermiaComponent}.
     */
```

File: src/main/java/org/terasology/climateConditions/HypothermiaSystem.java [hunk 6]
```diff
@@ -48,13 +53,19 @@ public void observeDangerZone(MovedEvent event, EntityRef player) {
...
     @ReceiveEvent(components = {HypothermiaComponent.class})
     public void modifySpeed(GetMaxSpeedEvent event, EntityRef player) {
         if (event.getMovementMode() == MovementMode.WALKING) {
             event.multiply(walkSpeedMultiplier);
         }
     }
 
+    /**
+     * Modifies the jump speed of player as long as the player has a Hypothermia Component.
+     */
```
Comment (jdrueckert) [inline src/main/java/org/terasology/climateConditions/HypothermiaSystem.java]:
```suggestion
    /**
     * Modifies the jump speed of the player.
     * Is only active iff the player has a {@link HypothermiaComponent}.
     */
```

File: src/main/java/org/terasology/climateConditions/VisibleBreathComponent.java [hunk 7]
```diff
@@ -3,6 +3,10 @@
 import org.terasology.entitySystem.Component;
 import org.terasology.entitySystem.entity.EntityRef;
 
+/**
+ * This component is added to the player entity when the player enters the danger zone. Gives a visible breath to the
+ * player using particle effects.
+ */
```
Comment (jdrueckert) [inline src/main/java/org/terasology/climateConditions/VisibleBreathComponent.java]:
```suggestion
/**
 * Augments the world with a visible breath particle effect.
 * Is added/updated by the {@link VisibleBreathingSystem} periodically.
 */
```

File: src/main/java/org/terasology/climateConditions/VisibleBreathingSystem.java [hunk 8]
```diff
@@ -31,6 +31,10 @@
 import org.terasology.particles.components.generators.VelocityRangeGeneratorComponent;
 import org.terasology.registry.In;
 
+/**
+ * This system is responsible for giving the player a visible breath using particle effects as long as the player has a
+ * hypothermia component.
+ */
```
Comment (jdrueckert) [inline src/main/java/org/terasology/climateConditions/VisibleBreathingSystem.java]:
```suggestion
/**
 * Adds a {@link VisibleBreathComponent} to the player.
 * Visible breath is a periodic action adding a visible breath particle effect.
 * Is active only iff the player has a {@link HypothermiaComponent}
 */
```

Actually when thinking about this and the "increases difficulty" part of the hypothermia component, I realized, that visible breath should be decoupled from hypothermia. Hypothermia is an "illness" that negatively affects a player - visible breath is independent of that and appears whenever it is cold enough, independent of whether or not the player has hypothermia.

**Review by jdrueckert — CHANGES_REQUESTED** [review]
Left a bunch of suggestions - not because your doc comments where bad in general, but because they felt less concise than they could be and because I didn't find a good way of how to generally describe ho to get them there :see_no_evil: 
Please tell me if in the future you'd rather get a "maybe rather go in that direction" instead of precise suggestions and I'll try to explain :sweat_smile:

**Comment by ktksan** [comment]
I'm sorry 😅😅 I wasn't really sure about the kind of comments in docs, so I just looked at the comments in BlockDamageComponent, etc.. and then wrote it like that.... I will try to make it concise next time.

File: src/main/java/org/terasology/climateConditions/HypothermiaSystem.java [hunk 9]
```diff
@@ -48,13 +52,21 @@ public void observeDangerZone(MovedEvent event, EntityRef player) {
         }
     }
 
+    /**
+     * Modifies the walking/running speed of the player.
```
Comment (skaldarnar) [inline src/main/java/org/terasology/climateConditions/HypothermiaSystem.java]: Well, that `modifySpeed` modifies the speed should be no surprise. However, in what way does it modify it? In particular, we are **decreasing** the **walk speed**.

File: src/main/java/org/terasology/climateConditions/HypothermiaSystem.java [hunk 10]
```diff
@@ -48,13 +52,21 @@ public void observeDangerZone(MovedEvent event, EntityRef player) {
...
+     */
     @ReceiveEvent(components = {HypothermiaComponent.class})
     public void modifySpeed(GetMaxSpeedEvent event, EntityRef player) {
         if (event.getMovementMode() == MovementMode.WALKING) {
             event.multiply(walkSpeedMultiplier);
         }
     }
 
+    /**
+     * Modifies the jump speed of the player.
```
Comment (skaldarnar) [inline src/main/java/org/terasology/climateConditions/HypothermiaSystem.java]: Same as above:  `modifyJumpSpeed` modifies the jump speed - no surprises there. But in what way? Another clue here might be the fact that we are not modifying any physical ability of the character but reacting to the respective events.

File: src/main/java/org/terasology/climateConditions/VisibleBreathComponent.java [hunk 11]
```diff
@@ -3,6 +3,10 @@
 import org.terasology.entitySystem.Component;
 import org.terasology.entitySystem.entity.EntityRef;
 
+/**
+ * Augments the world with a visible breath particle effect.
```
Comment (skaldarnar) [inline src/main/java/org/terasology/climateConditions/VisibleBreathComponent.java]: "Augments the world" sounds a bit ... broad and unspecific. I'd be a bit more specific and explain that this only has the desired effect if the entity also has a `LocationComponent` that determines the location and direction of the effect.

**Review by skaldarnar — COMMENTED** [review]
When writing documentation I try to look at the code as somebody who has a general understanding of the framework (e.g., knows what an entity or component is), but has no clue what the specific systems, etc. are doing. 

After writing down the first doc comments I go back and read them to me - do they give me more information than I had before? Do they comments raise more questions? 

A good example is "modifies" or "adjusts" - these verbs should be accompanied by an explanation, or better be replaced by more concrete wording. For instance, "modifies speed to be slower" could become "slows down", and instead of "adjust the modifier" you could clarify that "adjust speed modifier according to time spent in the danger zone".

**Review by jdrueckert — APPROVED** [review]

**Pull request merged**